pilot/pkg/proxy/envoy/v2/pushqueue.go

```diff
@@ -0,0 +1,83 @@
+package v2
+
+import (
+	"sync"
+
+	"istio.io/istio/pilot/pkg/model"
+)
+
+type PushInformation struct {
+	// If not empty, it is used to indicate the event is caused by a change in the clusters.
+	// Only EDS for the listed clusters will be sent.
+	edsUpdatedServices map[string]struct{}
+
+	push *model.PushContext
+
+	full bool
+}
+
+type PushQueue struct {
+	mu          sync.RWMutex
+	connections map[*XdsConnection]*PushInformation
+	order       []*XdsConnection
+	signal      chan struct{}
+}
+
+func NewPushQueue() *PushQueue {
+	return &PushQueue{
+		connections: make(map[*XdsConnection]*PushInformation),
+		signal:      make(chan struct{}),
+	}
+}
+
+func (p *PushQueue) Add(proxy *XdsConnection, pushInfo *PushInformation) {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+	info, exists := p.connections[proxy]
+	if !exists {
+		p.connections[proxy] = pushInfo
+		p.order = append(p.order, proxy)
+	} else {
+		if info.edsUpdatedServices == nil && len(pushInfo.edsUpdatedServices) != 0 {
+			info.edsUpdatedServices = map[string]struct{}{}
+		}
+		info.push = pushInfo.push
+		info.full = info.full || pushInfo.full
+		for endpoint := range pushInfo.edsUpdatedServices {
+			info.edsUpdatedServices[endpoint] = struct{}{}
+		}
+	}
+	select {
+	case p.signal <- struct{}{}:
+	default:
+	}
+}
+
+func (p *PushQueue) waitForPendingPush() {
+	p.mu.RLock()
+	pending := len(p.order)
+	if pending == 0 {
+		p.mu.RUnlock()
+		<-p.signal
+	} else {
+		p.mu.RUnlock()
+	}
+}
+
+func (p *PushQueue) Remove() (*XdsConnection, *PushInformation) {
+	p.waitForPendingPush()
+
```

Comment:
What is the purpose of this step?
The next statement will block until queue.add completed anyway.

In addition, there is a race between the 2 go routine s.
Next .add can begin when remove is about to execute mu.lock.

Reply:
The goal here is to have `Remove()` block until there is a change in the queue so we don't have to do something like
```go
for {
   proxy := Remove()
   if proxy == nil {
      sleep and try again
   }
}
````

If the sequence is Remove(), Add(), Add(), with the second Add being before mu.lock this is still fine I think -- this is pretty subtle though

I thought I had some comments around this, maybe they somehow got lost, I'll add some back

Reply:
race should be fixed now